pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright 2022 NVIDIA CORPORATION & AFFILIATES
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Mellanox/network-operator/api/v1alpha1"
+	"github.com/Mellanox/network-operator/pkg/consts"
+)
+
+type NodeUpgradeState struct {
+	Node            *v1.Node
+	DriverPod       *v1.Pod
+	DriverDaemonSet *appsv1.DaemonSet
+}
+
+type ClusterUpgradeState struct {
+	Policy *v1alpha1.OfedUpgradePolicySpec
+
+	NodeStates map[string][]*NodeUpgradeState
+}
+
+func NewClusterUpgradeState() ClusterUpgradeState {
+	return ClusterUpgradeState{NodeStates: make(map[string][]*NodeUpgradeState)}
+}
+
+type ClusterUpgradeStateManager struct {
+	K8sClient         client.Client
+	Log               logr.Logger
+	DrainManager      DrainManager
+	PodRestartManager PodRestartManager
+}
+
+func NewClusterUpdateStateManager(
+	drainManager DrainManager,
+	podRestartManager PodRestartManager,
+	log logr.Logger,
+	k8sClient client.Client) *ClusterUpgradeStateManager {
+	manager := &ClusterUpgradeStateManager{
+		DrainManager:      drainManager,
+		PodRestartManager: podRestartManager,
+		Log:               log,
+		K8sClient:         k8sClient,
+	}
+
+	return manager
+}
+
+func (m *ClusterUpgradeStateManager) ApplyState(ctx context.Context, currentState *ClusterUpgradeState) {
```

Comment:
This function should return error. Also, you should not ignore errors in ClusterUpgradeStateManager methods, for example when you failed to update node upgrade state in k8s API.